### ISSUE TITLE
password lookup doesn't always respect chars setting

### ISSUE THREAD

**Issue by cpatry-poly** [ansible/ansible#17461]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

password lookup
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/usr/share/ansible/']
```
##### CONFIGURATION

Pretty much default configuration
##### OS / ENVIRONMENT

Ansible server : Ubuntu 14.04.3 LTS
Managed Host : Debian 8.5
##### SUMMARY

Password don't respect specified characters and use punctuation when it shouldn't
##### STEPS TO REPRODUCE

It's a bit random. I have that issue frequently, but not each time.
When I try to reproduce the problem with a test playbook, I couldn't reproduce the issue.
May be I am crazy or something

```
- name: Create replication role
  postgresql_user:
    name: replication
    encrypted: yes
    password: "{{ lookup('password', '/etc/psql_credentials/' + ansible_hostname + ':replication chars=ascii_letters,digits') }}"
    role_attr_flags: LOGIN,REPLICATION
  become_user: postgres
```
##### EXPECTED RESULTS

The newly created file should contain a password with only ascii_letters and digits.
Something like that :

```
root@ans01:# cat /etc/psql_credentials/namindpsql01:replication
yn0P43VUTYsP4Q0kuZHz
root@ans01:# cat /etc/psql_credentials/alllndpsql01:replication
fgZpUPQtypsaCXkPmMb7
```
##### ACTUAL RESULTS

Instead I got this :

```
root@ans01:# cat /etc/psql_credentials/namindpsql01:replication
yn0P.3VUTYsP4Q0k.ZHz
root@ans01:# cat /etc/psql_credentials/alllndpsql01:replication
fgZpUPQ:-psaCXkPmMb7
```

Maybe it's not a bug and my password line is bad, but I can't figure out what I do wrong.
A little help will be much appreciated.

Best regards,

**Comment by alikins** [comment]
The playbook syntax looks okay to me. I added some unit test cases to add a similar line (with a colon in the filename, and a the same chars spec) and it seems to work okay.

Do the files '/etc/psql_credentials/namindpsql01:replication' already exist?
Does the content of the files change when ran. It shouldn't for that password lookup
line, as per the notes in http://docs.ansible.com/ansible/playbooks_lookups.html#the-password-lookup:

```
If the file already exists, no data will be written to it. If the file has contents, those contents will be read in as the password. Empty files cause the password to return as an empty string
```

**Comment by abadger** [comment]
@Tiduster I became aware of this through @alikins' pull request for adding unittests for this.  I'd seen a problem in the code which may be the same thing.  Started working on code to fix it here: https://github.com/ansible/ansible/pull/17475  I haven't tested the code yet, though, so there may be some  (even obvious) bugs in it but you're welcome to try it out if you'd like.  I'll talk to alikins tomorrow about combining our efforts.  Just at the end of a long long day right now and I'm eager for bed.

**Comment by abadger** [comment]
Actually, my fix is for non-ascii characters... looks like you're specifying all ascii and the output also contains ascii... it's just being drawn from a larger range than expected.  So it's likely that my code change won't fix this.

**Comment by cpatry-poly** [comment]
@alikins : Do the files '/etc/psql_credentials/namindpsql01:replication' already exist?

**No, it was created by the task.
I verified it my self using unix ctime and mtime.**

Does the content of the files change when ran. It shouldn't for that password lookup

**No, when the file is already there, the content never change.**

@abadger : thanks for letting me know.
Looking at the change you made, I saw something :
`
DEFAULT_PASSWORD_CHARS = to_text(ascii_letters + digits + ".,:-_")
`
.,:-_ are the only characters that appears when the lookup bug. Maybe they are always there, even if I specify ascii and digits only ?

**Comment by abadger** [comment]
@Tiduster That was my first guess too but so far I've been unable to reproduce.  I ran the following playbook 100 times with a checkout of v2.1.0.0-1

```
---
- hosts: localhost
  gather_facts: False
  tasks:
  - debug:
      msg: "{{ lookup('password', '/var/tmp/blah-' + inventory_hostname + ':replication chars=ascii_letters,digits') }}"
  - command: cat /var/tmp/blah-localhost:replication
    register: contents

  - file:
      path: "/var/tmp/blah-localhost:replication"
      state: absent
    when: "{{ '.' not in contents['stdout'] and ',' not in contents['stdout'] and ':' not in contents['stdout'] and '-' not in contents['stdout'] and '_' not in contents['stdout']}}"
```

for i in `seq 0 100` ; do ansible-playbook ../troubleshooting/17461.yml -vv ; done

The conditional never triggered to save the file so I know that there was never one of the punctuation characters from DEFAULT_PASSWORD_CHARS in there.

Are you able to run my playbook above and see if it generates passwords with the punctuation characters for you?  If it doesn't, perhaps there's another password lookup somewhere in your playbook that's generating the password before we get to this task?

**Comment by abadger** [comment]
needs_info

**Comment by cpatry-poly** [comment]
Hello !

I already tried this kind of test before and didn't reproduce the issue.
I ran your playbook directly on our ansible server and everything was fine.

So I looked again if I didn't do everything wrong before this task, as you ask.
I found a template which use this lookup inside : 
{{ lookup("password", '/etc/psql_credentials/' + master_hostname + ':replication') }}

And it was use before AND after the other lookup, so this explains everything.

Sorry for all the trouble :-( .

Case closed.

Best regards,
